crates/sui-core/src/transaction_driver/mod.rs

```diff
@@ -0,0 +1,256 @@
+    committee::EpochId,
+    effects::{TransactionEffects, TransactionEvents},
+    error::SuiError,
+    messages_grpc::SubmitTransactionRequest,
+    object::Object,
+    quorum_driver_types::{
+        EffectsFinalityInfo, ExecuteTransactionRequestV3, ExecuteTransactionResponseV3,
+        FinalizedEffects,
+    },
+};
+use tokio::{
+    task::JoinSet,
+    time::{sleep, timeout},
+};
+
+use crate::{
+    authority_aggregator::AuthorityAggregator,
+    authority_client::AuthorityAPI,
+    quorum_driver::{reconfig_observer::ReconfigObserver, AuthorityAggregatorUpdatable},
+};
+
+/// Options for submitting a transaction.
+#[derive(Clone, Default, Debug)]
+pub struct SubmitTransactionOptions {
+    /// When forwarding transactions on behalf of a client, this is the client's address
+    /// specified for ddos protection.
+    pub forwarded_client_addr: Option<SocketAddr>,
+}
+
+pub struct TransactionDriver<A: Clone> {
+    authority_aggregator: ArcSwap<AuthorityAggregator<A>>,
+    state: Mutex<State>,
+    metrics: Arc<TransactionDriverMetrics>,
+}
+
+impl<A> TransactionDriver<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    pub fn new(
+        authority_aggregator: Arc<AuthorityAggregator<A>>,
+        reconfig_observer: Arc<dyn ReconfigObserver<A> + Sync + Send>,
+        metrics: Arc<TransactionDriverMetrics>,
+    ) -> Arc<Self> {
+        let driver = Arc::new(Self {
+            authority_aggregator: ArcSwap::new(authority_aggregator),
+            state: Mutex::new(State::new()),
+            metrics,
+        });
+        driver.enable_reconfig(reconfig_observer);
+        driver
+    }
+
+    pub async fn submit_transaction(
+        &self,
+        // TODO(fastpath): Replace with TD request/response type once hooked up to transaction orchestrator
+        request: ExecuteTransactionRequestV3,
+        options: SubmitTransactionOptions,
+    ) -> Result<ExecuteTransactionResponseV3, TransactionDriverError> {
+        let tx_digest = request.transaction.digest();
+        let is_single_writer_tx = !request.transaction.contains_shared_object();
+        let timer = Instant::now();
+        loop {
+            match self.submit_transaction_once(&request, &options).await {
+                Ok(resp) => {
+                    let settlement_finality_latency = timer.elapsed().as_secs_f64();
+                    self.metrics
+                        .settlement_finality_latency
+                        .with_label_values(&[if is_single_writer_tx {
+                            TX_TYPE_SINGLE_WRITER_TX
+                        } else {
+                            TX_TYPE_SHARED_OBJ_TX
+                        }])
+                        .observe(settlement_finality_latency);
+                    return Ok(resp);
+                }
+                Err(e) => {
+                    tracing::warn!("Failed to submit transaction {tx_digest}: {}", e);
+                    sleep(Duration::from_secs(1)).await;
```

Comment:
We probably have past data that could help us reason about the expected avg number of retries that will be needed? Could we dig that up?

Reply:
It doesn't look like we retry that often based on mainnet results. I know we removed the conflicting tx retries but I would still have expected mroe. Does this check out with you?

https://metrics.sui.io/goto/PBK1TToHR?orgId=1

Reply:
Is that the quorum-level retry metric? That probably isn't useful here since we are picking a random validator. So the retry rate will be the probability that it might pick a validator that would fail to return the right response?

Reply:
Good call, checking the authority agg metrics we get a better picture. I have filtered down to the retryable errors - https://metrics.sui.io/goto/rjrLsTTHg?orgId=1

We will to build out heuristics for selecting which validator to submit with but once we have our set of "good" validators I think we will have to continue to retry a tx until we have exhausted that set of validators. And for each validator we can stop retrying after a non-retryable error is returned and then based on the metrics for retryable error rate maybe 2-3 retries is enough per validator

Reply:
Would it trigger DoS if it submits to a few validators at the same time? Or is there a threshold (e.g. cannot duplicate for more than K or something)

Reply:
Last we spoke about it, this should be configurable as part of the DoS protection mechanism but @mwtian can correct me if I am wrong

Reply:
If the error is because of a timeout, then adding an additional sleep doesn't make much sense here.

Reply:
If a validator does not acknowledge transaction submission, then the transaction driver should notice this and not submit through that validator when possible. If DoS protection is triggered, it is temporary and eventually social slashing can happen against the bad validator.

Reply:
We can add a TODO to handle errors differently by types. Right now, a common case in tests is that the full node spins against down validators. So some delay is needed at least after exhausting all possible peers.